File: server/typescript/src/application/use-cases/update-item.use-case.ts
import type { IItemRepository } from '../../domain/repositories/item.repository.port.js';
import { NotFoundError, ValidationError } from '../../shared/errors.js';
import { ValidationService } from '../services/validation.service.js';
import type { ReplaceItemData } from '../../domain/entities/item.entity.js';
import type { Item } from '../../domain/entities/index.js';
import { Condition, Status } from '../../domain/entities/index.js';

/**
 * Update item (PUT - full replace) use case
 */
export class UpdateItemUseCase {
  constructor(private readonly itemRepository: IItemRepository) {}

  /**
   * Execute the use case
   */
  async execute(id: number, data: ReplaceItemData): Promise<Item> {
    // Validate input
    const validatedData = ValidationService.validateUpdateItem(data);

    // Check if item exists
    const exists = await this.itemRepository.exists(id);
    if (!exists) {
      throw new NotFoundError('Item', id);
    }

    // Cast enums and update item in repository
    const updateData = {
      ...validatedData,
      condition: validatedData.condition as Condition,
      status: validatedData.status as Status
    };

    const updatedItem = await this.itemRepository.update(id, updateData);

    if (!updatedItem) {
      throw new NotFoundError('Item', id);
    }

    return updatedItem;
  }
}

File: server/kotlin/src/main/kotlin/com/lebonpoint/application/usecases/UpdateItemUseCase.kt
package com.lebonpoint.application.usecases

import com.lebonpoint.application.services.ValidationService
import com.lebonpoint.domain.entities.Item
import com.lebonpoint.domain.entities.ReplaceItemData
import com.lebonpoint.domain.repositories.ItemRepository
import com.lebonpoint.shared.NotFoundException

/**
 * Use case for updating (replacing) an item
 */
class UpdateItemUseCase(
    private val repository: ItemRepository
) {
    /**
     * Execute the use case
     * @param id Item ID
     * @param data Item replacement data
     * @return Updated item
     * @throws NotFoundException if item not found
     * @throws ValidationException if validation fails
     */
    suspend fun execute(id: Int, data: ReplaceItemData): Item {
        ValidationService.validateId(id)
        ValidationService.validateTitle(data.title)
        ValidationService.validatePriceCents(data.priceCents)
        ValidationService.validateCondition(data.condition)
        ValidationService.validateStatus(data.status)

        return repository.update(id, data)
            ?: throw NotFoundException("Item", id)
    }
}
